### ISSUE TITLE
Interpolated fields defined in config not triggering ForceNew

### ISSUE THREAD

**Issue by rileykarson** [hashicorp/terraform-plugin-sdk#122]
### Terraform Version

```
 v  github.com/hashicorp/terraform/helper/config  v0.11.2  v0.11.2
```

### Terraform Configuration Files

```diff
resource "google_storage_bucket" "org_public" {
  name          = "org_public"
  location      = "US"
  storage_class = "MULTI_REGIONAL"
}

resource "google_storage_bucket_object" "bootstrap_linux" {
  name   = "bootstrap_linux.sh"
-  source = "./scripts/bootstrap_linux.sh"
+  source = "someOtherFile"
  bucket = "${google_storage_bucket.org_public.name}"
}

resource "google_storage_object_acl" "bootstrap_linux-acl" {
  bucket = "${google_storage_bucket.org_public.name}"
  object = "${google_storage_bucket_object.bootstrap_linux.name}"

  predefined_acl = "publicRead"
}
```

### Debug Output

N/A

### Crash Output

N/A

### Expected Behavior

`google_storage_object_acl` recreates, is applied to a new `google_storage_object`

### Actual Behavior

Only `google_storage_object` is recreated.

```bash
data.terraform_remote_state.suqld_core_state: Refreshing state...
google_storage_bucket.org_public: Refreshing state... (ID: org_public)
google_storage_bucket_object.bootstrap_linux: Refreshing state... (ID: org_public-bootstrap_linux.sh)
google_storage_object_acl.bootstrap_linux-acl: Refreshing state... (ID: bootstrap_linux.sh-acl)

An execution plan has been generated and is shown below.
Resource actions are indicated with the following symbols:
-/+ destroy and then create replacement

Terraform will perform the following actions:

-/+ google_storage_bucket_object.bootstrap_linux (new resource required)
      id:             "org_public-bootstrap_linux.sh" => <computed> (forces new resource)
      bucket:         "org_public" => "org_public"
      content_type:   "text/plain; charset=utf-8" => <computed>
      crc32c:         "FshAmA==" => <computed>
      detect_md5hash: "YTxGlR/gtM3nOD1QZG+QxA==" => "different hash" (forces new resource)
      md5hash:        "YTxGlR/gtM3nOD1QZG+QxA==" => <computed>
      name:           "bootstrap_linux.sh" => "bootstrap_linux.sh"
      source:         "../scripts/bootstrap_linux.sh" => "../scripts/bootstrap_linux.sh"
      storage_class:  "MULTI_REGIONAL" => <computed>


Plan: 1 to add, 0 to change, 1 to destroy.
```

Terraform doesn't recreate the downstream resource because `name` is a `Required` field, and the user has it set - so it doesn't see a change. If this resource used a computed `id`/`self_link`, Terraform would see a change from `aValue => <computed>` and recreate.

Terraform should recreate downstream resources that interpolate from a field in a user's config when the upstream resource is recreated.

### Steps to Reproduce

```
terraform apply
```

### References

https://github.com/terraform-providers/terraform-provider-google/issues/2213

**Comment by radeksimko** [comment]
Thank you @rileykarson for the config - I managed to reproduce it in Terraform `0.12.18` & GCP provider `3.1.0`.

It is reasonable to expect Terraform to recreate the object ACL when the object itself is recreated.
Unfortunately though it is not a behaviour we can implement and just generically apply to all resources. There are cases where the current behaviour makes more sense.

For example the user could be referencing `google_storage_bucket_object.bootstrap_linux.name` in [`google_compute_instance_template.metadata_startup_script`](https://www.terraform.io/docs/providers/google/r/compute_instance_template.html#metadata_startup_script) in which case I'm pretty sure they wouldn't want to recreate the template and any instances or groups of instances which are based on the template.

I assume the SDK here somehow needs to allow providers to differentiate between the two different ways resources depend on each other and pass that to core so it can use that information accordingly when building the graph.

**Comment by rileykarson** [comment]
For what it's worth, that would be my expectation in that case. [That field](https://github.com/terraform-providers/terraform-provider-google/blob/master/google/resource_compute_instance_template.go#L223-L227) is marked `ForceNew`, and changing the underlying script recreating the template (and ergo any resources with hard dependencies on the template) makes sense.

Reading [Martin's comment in #113](https://github.com/hashicorp/terraform-plugin-sdk/issues/113#issuecomment-412701726), specifically this bit, I'm surprised:

> Terraform Core is actually behaving as currently designed here because dependencies are used only to decide ordering, not to create additonal actions in the plan.

I would have expected new actions in plan to be taken in these cases, and some of the TF behaviour that I've considered a bug makes more sense in that context.

---

However, as a provider developer, I've probably got different expectations than many users. I'm *sure* there's a counterexample to my expectations, but I haven't been able to think of any concrete ones yet.

**Comment by jhult** [comment]
Any news on this?

P.S. I appreciate those of you who are contributing code to this repository!

**Comment by bflad** [comment]
The terraform-plugin-sdk/v2 behavior of not triggering resource replacement via `ForceNew` is not likely to ever change in this regard as the underlying value it is receiving in the plan, as compared to the prior state, is not changing. The protocol does not provide any additional information about the source of a value (e.g. that the value came a reference of another resource and that said resource is being replaced) so there is no other method to determine that this type of behavior is necessary from the SDK's perspective. I would venture a guess that designing that type of additional functionality in the protocol would require a lot of effort and would require prioritization over other feature work if deemed acceptable with Terraform's design principles.

On the other side of the protocol, the Terraform CLI behavior of not propagating resource replacements through all references of a resource being replaced will likely not change due to compatibility promises and lack of context when such a propagation should occur, since not all downstream resources may warrant this behavior. It could be very destructive.

There is, however, a recent internal proposal in the works which would allow practitioners to declare arbitrary resource replacement references via the Terraform configuration. In that proposal, other any whole resource or attribute references can be used to replace the resource without affecting the remote resource configuration (e.g. adding this Terraform-only configuration detail to trigger resource replacement when there is an updated parent resource id or timestamp value). This will likely be a recommended solution for provider documentation going forward if/when it is implemented. I'm unable to provide additional details though as the design and prioritization of that effort is still ongoing.

**Comment by Chupaka** [comment]
@bflad,
> other any whole resource or attribute references can be used to replace the resource without affecting the remote resource configuration (e.g. adding this Terraform-only configuration detail to trigger resource replacement when there is an updated parent resource id or timestamp value)

Like "keepers" in random_id?

**Comment by bflad** [comment]
@Chupaka correct. It'd be a generic configuration implementation that each resource would get without change. Essentially it would supersede and replace implementations such as [`keepers` in the random provider](https://registry.terraform.io/providers/hashicorp/random/latest/docs#resource-keepers) and `triggers` in the [null](https://registry.terraform.io/providers/hashicorp/null/latest/docs/resources/resource#triggers) and [time](https://registry.terraform.io/providers/hashicorp/time/latest/docs#resource-triggers) providers.

**Comment by memphis88** [comment]
I just stumbled on a situation similar to https://github.com/hashicorp/terraform-provider-google/issues/7235 which as I understand is relevant to the topic here. I understand that the topic is more elaborate than one would think from the get-go but I'm curious if there are any workarounds for the time being since I cannot seem to find any.

**Comment by bflad** [comment]
Hi Everyone 👋 

[Terraform CLI 1.2.0-beta1](https://github.com/hashicorp/terraform/releases/tag/v1.2.0-beta1) now includes the ability to declare arbitrary resource replacement references via the Terraform configuration [`lifecycle` configuration block `replace_triggered_by` argument](https://github.com/hashicorp/terraform/pull/30931/files) (beta documentation is not present on terraform.io at the moment). This can be used with any resource. other any whole resource or attribute references can be used to replace the resource without affecting the remote resource configuration (e.g. adding this Terraform-only configuration detail to trigger resource replacement when there is an updated parent resource id or timestamp value). This will likely be a recommended solution for provider documentation going forward if/when it is implemented.

If you have any questions about the Terraform CLI feature, please reach out in the [HashiCorp Discuss Forums](https://discuss.hashicorp.com/c/terraform-core/27). If you have any bug reports or enhancement requests with the feature, please reach out by searching for or creating an issue in the [Terraform CLI GitHub Repository](https://github.com/hashicorp/terraform/issues). Thanks!

**Comment by bflad** [comment]
Please note that [Terraform CLI 1.2.0 has been released today](https://www.hashicorp.com/blog/terraform-1-2-improves-exception-handling-and-updates-to-the-cli-driven-workflow), and [the documentation for this feature is now live on terraform.io](https://www.terraform.io/language/meta-arguments/lifecycle#replace_triggered_by).
